modules/caddyhttp/app.go

```diff
@@ -493,31 +497,31 @@ func (app *App) Start() error {
 func (app *App) Stop() error {
 	ctx := context.Background()
 
-	// see if any listeners in our config will be closing or if they are continuing
-	// hrough a reload; because if any are closing, we will enforce shutdown delay
+	// decrement the HTTP app reference count; get the current count
+	// so that we can detect whether this was the last reference,
+	// meaning that this is a shutdown and not a config reload.
+	currentRefCount := atomic.AddInt32(&appRefCount, -1)
+
+	// set up the shutdown delay if this is the last HTTP app instance running
 	var delay bool
 	scheduledTime := time.Now().Add(time.Duration(app.ShutdownDelay))
 	if app.ShutdownDelay > 0 {
 		for _, server := range app.Servers {
-			for _, na := range server.addresses {
-				for _, addr := range na.Expand() {
-					if caddy.ListenerUsage(addr.Network, addr.JoinHostPort(0)) < 2 {
-						app.logger.Debug("listener closing and shutdown delay is configured", zap.String("address", addr.String()))
-						server.shutdownAtMu.Lock()
-						server.shutdownAt = scheduledTime
-						server.shutdownAtMu.Unlock()
-						delay = true
-					} else {
-						app.logger.Debug("shutdown delay configured but listener will remain open", zap.String("address", addr.String()))
-					}
-				}
+			if currentRefCount == 0 {
```

Comment:
The previous `shutdown_delay` used to be scoped to each 
`app.Servers`'s `addr.Network, addr.JoinHostPort(0)` combination (`lnKey`) in the `listenerPool` `UsagePool`.

Meaning, previously, a config with multiple `apps.http.servers` would always engage `shutdown_delay`, if a single of those got removed (or a new one added).

This no longer happens. It's global now, only triggering `if currentRefCount == 0 {`.

Not sure which version makes more sense.

---

I dug a bit more around and found https://github.com/caddyserver/caddy/pull/4906#issuecomment-1201841104 from the original PR

> [...] Now shutdowns are announced per-server and whether it's a process exit or a reload is irrelevant.

Following a commit (and refactor) explicitly making it per-server /shrug

Reply:
Right... this is technically a behaviour change from the original intent, I think :frowning_face: 

@mholt so you see value in retaining that per-address shutdown delay (i.e. delay reloads if a server is actually going away)? I think this is extremely rare to happen in practice, how often do people remove a listener and keep the other? And what are the chances they care about delaying in that case?

I feel like the shutdown delay only really has value if Caddy is actually shutting down i.e. for a re-deployment or process restart so that some fronting load-balancer can notice before it goes down.

Reply:
It's common. I think we need to keep ListenerUsage() accurate. I'll try to look into this soon and see if I can figure out a fix that does this.